### ISSUE TITLE
Error: cannot cast type boolean to jsonb

### ISSUE THREAD

**Issue by uasan** [porsager/postgres#386]
Hello.
Version `postgres 3.2.4`.

```JS
await sql`SELECT ${true}::jsonb`;
```

Result postgresql error:
```
Cannot cast type boolean to jsonb
```

**Comment by Minigugus** [comment]
Maybe an issue with parameters types inference :thinking: On the meantime, maybe you can use `sql.json` instead?
```ts
await sql`SELECT ${sql.json(true)}::jsonb`;
```

**Comment by Minigugus** [comment]
> I do not know the type of the parameter

Ok but your cast to `::jsonb` will make `postgres` call `JSON.stringify` anyway, so where is the problem? `sql.json` just returns an object that tells `postgres` which oid to use in the query, so if you always cast to JSONB it should work I guess

**Comment by porsager** [comment]
@uasan oh shoot, that's actually not the only issue. Same goes for the other inferred types (Date, Buffer, Bigint) :-( I should have caught this for the v3 update and completely removed inferType (https://github.com/porsager/postgres/blob/218a7d4f37dcbf76d01081413a386c00544ab63e/src/types.js#L213) to let PostgreSQL decide :-/ Changing this would require a major bump, but my initial idea right now is that it would be the best direction.

**Comment by uasan** [comment]
You, like many good developers, simply do not have enough tests for many cases, you need to think about how the community can help you write tests

**Comment by uasan** [comment]
I think the best behavior in cases of `unknown` data type is to do a direct mapping of JS types to Postgres types, i.e.
```
JS boolean -> Postgres boolean
JS number -> Postgres numeric
JS bigInt -> Postgres numeric
```

**Comment by AkeleyUA** [comment]
@porsager Hi. Should we expect a fix soon? If yes, how long?

P.S. I don't want to raise, but this is a pretty critical issue for us. Wish you happy coding.

**Comment by porsager** [comment]
@AkeleyUA the issue is that we can't really change it without potentially breaking someone who relies on the implicit inference. It would require a new major, and I haven't thought through if that's the way to go yet. Could you maybe describe your specific issue / show an example query, and we can see if there's a way to work around it?

**Comment by uasan** [comment]
> show an example query, and we can see if there's a way to work around it?

Dynamic query builder that knows the column name and values, but does not know the type of the column
```JS
sql` UPDATE SET "${sql(name)}" = ${value}`
```

> potentially breaking someone who relies on the implicit inference

I'm pretty sure no one is relying on ghost `true` booleans when `json` is expected, I'm pretty sure this bug is not reproducible now because everyone is using the old `sql.json()` helper interface, I don't think fixing this bug will break backwards compatibility.

**Comment by porsager** [comment]
I think It's useful when calling functions that don't do the same implicit type casts as regular queries. I'm pretty sure someones setup will break if changing it, and the fact that it's an explicit functionality tested for, it would definitely require a major bump (but I don't mind if that's necessary).

So if we make that change, it would be removing all implicit type inference from js types supplied in the Parse message and rely completely on the PostgreSQL `ParameterDescription`. The actual code fix could just be changing the `inferType` function to `x => x instanceof Parameter ? x.type : 0`.

I'll try to see if I can find any good reasons not to do it 😉

**Comment by uasan** [comment]
the inference data type for Postgres from the data type of JS, has sense only if the `ParameterDescription` returned an unknown type, if I'm wrong, give an example that proves you are right, I don't argue, I'm even curious)

**Comment by porsager** [comment]
Yeah I'm very curious too ☺️

I think it's not possible because the types specified from our end are supplied in the `Parse` message which comes before the `ParameterDescription` message. So there is no way for us to tell Postgres "ok you don't know the type so we would actually like to use x".

**Comment by porsager** [comment]
I mean at least not in an efficient way, that would require us to call `Parse` first with unknown types, and then create a new extended query calling `Parse` again with our updated types after checking for unknown types in `ParameterDescription`.

**Comment by uasan** [comment]
> calling Parse again

I understand that you do not need to call `PARSE` again, you call it only once, at the `DESCRIBE` stage you also call it only once, so you created `STATEMENT` and you can call `BIND` many times without calling `PARSE` and `DESCRIBE` again

**Comment by porsager** [comment]
Right, but we can't supply types to `Bind`?

**Comment by uasan** [comment]
> but we can't supply types to Bind?

Send a one-time `PARSE` and `DESCRIBE` message, you receive a response types for all parameters, i.e. at the `BIND` call stage, you always already have all the known types

**Comment by uasan** [comment]
Here are links to my unfinished postgres client:

One-time creation of `STATEMENT`
https://github.com/uasan/postgres/blob/master/src/request/statement.js#L51-L57

Reusable type reuse in a BIND call (without re-call `PARSE` and `DESCRIBE`)
https://github.com/uasan/postgres/blob/master/src/request/statement.js#L65

**Comment by uasan** [comment]
You can pass alternative types to `BIND` when you see fit, I just want to say that if `DESCRIBE` returned a type, you need to use only that type without type inference.

**Comment by porsager** [comment]
No you can only decide in bind if a parameter is text or binary encoded. That's not related to postgres types. 
The types are static for the statement.

**Comment by uasan** [comment]
Yes, of course, during the life of a `STATEMENT`, you cannot change the types of parameters, but this is also not necessary for anyone or not? )

**Comment by tantaman** [comment]
Inserts like this (a boolean inside of an array):

```sql
INSERT INTO foo (a) VALUES (${[false]})
```

also fail with `column "a" is of type jsonb but expression is of type boolean`  which.. is surprising.
